src/combat.cpp

```diff
@@ -731,7 +876,54 @@ void Combat::CombatFunc(Creature* caster, const Position& pos, const AreaCombat*
 				}
 
 				if (!params.aggressive || (caster != creature && Combat::canDoCombat(caster, creature) == RETURNVALUE_NOERROR)) {
-					func(caster, creature, params, data);
+					CombatDamage damageCopy = damage; // we cannot avoid copying here, because we don't know if it's player combat or not, so we can't modify the initial damage.
+					bool playerCombatReduced = false;
+					if ((damageCopy.primary.value < 0 || damageCopy.secondary.value < 0) && caster) {
+						Player* targetPlayer = creature->getPlayer();
+						if (targetPlayer && caster->getPlayer() && targetPlayer->getSkull() != SKULL_BLACK) {
+							damageCopy.primary.value /= 2;
+							damageCopy.secondary.value /= 2;
+							playerCombatReduced = true;
+						}
+					}
+
+					damageCopy.primary.value += playerCombatReduced ? criticalPrimary / 2 : criticalPrimary;
+					damageCopy.secondary.value += playerCombatReduced ? criticalSecondary / 2 : criticalSecondary;
+
+					if (damageCopy.critical) {
+						g_game.addMagicEffect(creature->getPosition(), CONST_ME_CRITICAL_DAMAGE);
+					}
+
+					bool success = false;
+					if (damageCopy.primary.type != COMBAT_MANADRAIN) {
+						if (g_game.combatBlockHit(damageCopy, caster, creature, params.blockedByShield, params.blockedByArmor, params.itemId != 0)) {
+							continue;
+						}
+						success = g_game.combatChangeHealth(caster, creature, damageCopy);
+					} else {
+						success = g_game.combatChangeMana(caster, creature, damageCopy);
+					}
+
+					if (success) {
+						if (params.origin != ORIGIN_MELEE && damageCopy.primary.value != 0 && damageCopy.secondary.value != 0) {
+							for (const auto& condition : params.conditionList) {
+								if (caster == creature || !creature->isImmune(condition->getType())) {
+									Condition* conditionCopy = condition->clone();
+									conditionCopy->setParam(CONDITION_PARAM_OWNER, caster->getID());
+
+									//TODO: infight condition until all aggressive conditions has ended
```

Comment:
Is there something missing?

Reply:
If you mean that „todo” then it’s comment from the base tfs code

Reply:
@DSpeichert 
This should of been merged ages ago.
Leech amounts are busted without it.

Reply:
I can probably add later that todo, wont put all of it in one pr.